### PR TITLE
set correct labelmap for rknn detector

### PR DESCRIPTION
**Pull request by MarcA711**
As discussed [here](https://github.com/blakeblackshear/frigate/discussions/11502#discussioncomment-9582338), the labelmap for the models provided by frigate's rknn detector are wrong.
I tried to implement a function to load the correct (ms-coco) labelmap, but it doesn't work. I played a video of a cat (id 15 in yolonas and in the new `labelmap-coco.txt` that I try to load) but Frigate show it is a bird (which has id 15 in default labelmap.txt). I checked:
- the detector really returns id 15
- the new function `load_labelmap` that I implemented gets called correctly
- `load_labels` inside the `load_labelmap` function return the correct dictionary with cat as ID 15 (and not 16 like in the default labelmap)
- `self.labelmap` inside the `load_labelmap` function is empty `{}` (so it doesn't override anything)

Does anybody have an idea why it still says it is a bird? I see no difference to the way Frigate+ overrides the labelmap. Is it working correctly for Frigate+?

@blakeblackshear I think the way I implemented this function is the way the user expects it to. The correct labelmap is loaded if the user doesn't change the labelmap_path, if he sets a custom labelmap_path it won't be changed. I even explained it in detail in the docs. I think this will work for everyone. If you still has concerns, please tell me.

### CODE REVIEW

**Comment by blakeblackshear** [comment]
Where did you get this labelmap from? The default labelmap for COCO has 91 classes, but this one only has 80. Is this a YOLO-NAS specific labelmap because they exclude some classes during training? Or is this some kind of change that happens as a part of converting to rknn?

**Comment by blakeblackshear** [comment]
I would like to handle this in a way that is not detector specific given this draft PR I put up today: https://github.com/blakeblackshear/frigate/pull/11645

**Comment by MarcA711** [comment]
Hey @blakeblackshear 

> Where did you get this labelmap from? The default labelmap for COCO has 91 classes, but this one only has 80. Is this a YOLO-NAS specific labelmap because they exclude some classes during training? Or is this some kind of change that happens as a part of converting to rknn?

You are right, it seems coco also has a version with 91 classes. I used [this labelmap](https://docs.ultralytics.com/datasets/detect/coco/#dataset-yaml) from ultralytics from the time where I used yolov8. I kind of assumed that yolonas uses the same. I will do some research which labelmap is the correct.
AFAIK, during the conversion to rknn the classes shouldn't change.

> I would like to handle this in a way that is not detector specific given this draft PR I put up today: https://github.com/blakeblackshear/frigate/pull/11645

I actually included `load_labelmap` as a member function of `ModelConfig`, see https://github.com/MarcA711/frigate-rockchip/blob/ec4751bcb0d6f3f739a1d7aaa5c92cf038af1e82/frigate/detectors/detector_config.py#L136-L140. So, the code is not detector specific and can be used by any detector. However, I included the new labelmap in the rockchip image for now, because I wasn't sure if other images need this exact labelmap. I will include it in the main docker image directly, if you need it as well.

**Comment by MarcA711** [comment]
From [this line](https://github.com/Deci-AI/super-gradients/blob/217353a3a2fc919d2330d9c74f78e40d96d72b35/src/super_gradients/training/pretrained_models.py#L69) I think that yolonas uses 80 classes coco as well. Maybe they excluded 11 uncommon classes for better performance.

**Comment by MarcA711** [comment]
I did some additional research.

The problem is that `init_config` in `frigate/app.py` creates a new config based on the `config.yml` in:
https://github.com/blakeblackshear/frigate/blob/8c325801efba40c3a8f0f6deb28093942e7b20ba/frigate/app.py#L140
This new `self.config` doesn't have the updated labelmap.
Afterwards, the function `start_camera_processors` passes the labelmap based on this new config:
https://github.com/blakeblackshear/frigate/blob/8c325801efba40c3a8f0f6deb28093942e7b20ba/frigate/app.py#L493

Based on this observation, I wonder how frigate plus updates the labelmap in this function `check_and_load_plus_model` from `frigate/detectors/detector_config.py`:
https://github.com/blakeblackshear/frigate/blob/8c325801efba40c3a8f0f6deb28093942e7b20ba/frigate/detectors/detector_config.py#L114-L117
Does this have something to do with `user_config.runtime_config(self.plus_api)` from above? Can this be generalized to work for all detectors?

**Comment by blakeblackshear** [comment]
We can't assume this is the right labelmap just because they are using yolonas or rknn. This will be the wrong labelmap if Frigate+ or any other custom models are trained on yolonas. Also, no other models attempt to automatically choose a modified labelmap.

So here is what I think we should do, all in the main Dockerfile:

1. Move the embedded labelmaps into `/labelmaps/`, but leave the current one at root for backwards compatability
2. Put `/labelmaps/coco-complete.txt` and `/labelmaps/coco-yolo-80.txt` in place
3. Remove the code to automatically load the labelmap
4. Update the docs to direct users to use `/labelmaps/coco-yolo-80.txt` when using the yolonas model with pretrained weights.

**Comment by MarcA711** [comment]
I don't load the labelmap based on the model type. I load the labelmap only if the user uses one of the models that I created e.g. `model: path: deci-fp16-yolonas_s`
This will always be this model from my repo: https://github.com/MarcA711/rknn-models/releases/tag/v2.0.0

If the user uses a custom model like `model: path: /path/to/my_yolonas.rknn` I will not change the labelmap. So, if he uses a frigate+ model there will be no difference.

Also, even if the user uses the model I created (e.g. `model: path: deci-fp16-yolonas_s`) he can still overwrite the labelmap by using `model: labelmap_path` . If he sets this option, I will also not change the labelmap and the user can use his own labelmap even if he uses one of the models that I created.

I personally think if the user uses one of the models that Frigate loads, the correct labelmap should be loaded as well. The user shouldn't need to look up the correct labelmap path within the frigate image. However, if he provides his own model, he is responsible for loading it.

I think we should integrate similar methods for other detectors as well, because this is more user-friendly.

**Comment by blakeblackshear** [comment]
I don't disagree that it should work that way in the future. I know why this works for plus models and not here, but I don't think this is the PR to fix/overhaul that. For now, I think this should work just like the docs for using openvino with the bundled model. The docs show the config for the setting the model path and the labelmap path.

**Comment by blakeblackshear** [comment]
We ran into the same issue with Frigate+ model paths. If you set the model path (`plus://...`) nested under the detector, it doesn't work. It must be set at the root level.

**Comment by MarcA711** [comment]
Alright, I will update the docs then.

**Comment by MarcA711** [comment]
Yes, I had a look at how frigate+ does this and I try to implement a way for all detectors to load the correct labelmap, if I have some spare time in the future.

**Comment by MarcA711** [comment]
@blakeblackshear I couldn't stop thinking about it and kept looking at the code and understand it better now and know why the approach in this PR doesn't work.

I have an idea how to implement the feature with just a few minor changes. Every detector could (but doesn't have to) implement a function to load a labelmap for bundled models. If the user provides his own labelmap, it will be used and will not be overwritten.

Should I try to implement the idea or is there currently no interest in implementing this feature?

**Comment by blakeblackshear** [comment]
@MarcA711 I think what you suggested sounds fine. The code around detectors and models has gotten a bit messy, so I'm hoping to cleanup the interfaces a bit for consistency when I touch them.

At this time, the only way to use multiple detectors is if they use the same model. I would like to cleanup the detector config so that the only model related config option is the path. Everything else should be specified under the model at the top level config. This would enable you to run tensorrt, openvino, etc if they all support the same model. I think the labelmap should be specified under the model rather than the detector since it should be the same regardless of detector.

Apologies if the above isn't that clear, but it's a general idea of what I am thinking.

**Comment by MarcA711** [comment]
This sounds great!
